Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler, StandardScaler, OneHotEncoder, LabelEncoder


Load Data

In [2]:
sales_df = pd.read_csv('../Data/Sales.csv')
items_df = pd.read_csv('../Data/item.csv')
promotion_df = pd.read_csv('../Data/Promotion.csv')
supermarkets_df = pd.read_csv('../Data/Supermarkets.csv')

In [3]:
sales_df.head()

code  amount  units  time  province  week  customerId  supermarket  \
0  7680850106    0.80      1  1100         2     1      125434          244   
1  3620000470    3.59      1  1100         2     1      125434          244   
2  1800028064    2.25      1  1137         2     1      108320          244   
3  9999985067    0.85      1  1148         2     1      162016          244   
4  9999985131    2.19      1  1323         2     1       89437          244   

   basket  day  voucher  
0       1    1        0  
1       1    1        0  
2       2    1        0  
3       3    1        0  
4       4    1        0

In [4]:
items_df.head()

code                     descrption    type        brand      size
0  3000005040  AUNT JEM ORIGINAL PANCAKE MIX  Type 1  Aunt Jemima      2 LB
1  3000005070      A/JEM COMPLETE PANCAKE MI  Type 1  Aunt Jemima  32    OZ
2  3000005300      AJ BUTTERMILK PANCAKE MIX  Type 1  Aunt Jemima     32 OZ
3  3000005350       A J BTRMLK COMP PNCK MIX  Type 1  Aunt Jemima      1 LB
4  1600015760      BC PANCAKE MIX BUTTERMILK  Type 1     Bisquick   6.75 OZ

In [5]:
promotion_df.head()

code  supermarkets  week                feature            display  \
0  2700042240           285    91         Not on Feature  Mid-Aisle End Cap   
1  2700042292           285    92  Interior Page Feature     Not on Display   
2  2700042274           285    92  Interior Page Feature     Not on Display   
3  2700042273           285    92  Interior Page Feature     Not on Display   
4  2700042254           285    92  Interior Page Feature     Not on Display   

   province  
0         2  
1         2  
2         2  
3         2  
4         2

In [6]:
supermarkets_df.head()

supermarket_No  postal-code
0             199        30319
1             200        30134
2             201        30066
3             202        31093
4             203        30542

Data Exploration

In [7]:
#Inspect the data structure
for df_name, df in zip(['Sales', 'Items', 'Promotion', 'Supermarkets'], 
                       [sales_df, items_df, promotion_df, supermarkets_df]):
    print(f"\n{df_name} DataFrame:")
    print(df.info())
    print("\nFirst few rows:")
    print(df.head())


Sales DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 11 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   code         1048575 non-null  int64  
 1   amount       1048575 non-null  float64
 2   units        1048575 non-null  int64  
 3   time         1048575 non-null  int64  
 4   province     1048575 non-null  int64  
 5   week         1048575 non-null  int64  
 6   customerId   1048575 non-null  int64  
 7   supermarket  1048575 non-null  int64  
 8   basket       1048575 non-null  int64  
 9   day          1048575 non-null  int64  
 10  voucher      1048575 non-null  int64  
dtypes: float64(1), int64(10)
memory usage: 88.0 MB
None

First few rows:
         code  amount  units  time  province  week  customerId  supermarket  \
0  7680850106    0.80      1  1100         2     1      125434          244   
1  3620000470    3.59      1  1100         2     1      125434   

In [8]:
# Check for missing values
for df_name, df in zip(['Sales', 'Items', 'Promotion', 'Supermarkets'], 
                       [sales_df, items_df, promotion_df, supermarkets_df]):
    print(f"\nMissing values in {df_name} DataFrame:")
    print(df.isnull().sum())


Missing values in Sales DataFrame:
code           0
amount         0
units          0
time           0
province       0
week           0
customerId     0
supermarket    0
basket         0
day            0
voucher        0
dtype: int64

Missing values in Items DataFrame:
code          0
descrption    0
type          0
brand         0
size          0
dtype: int64

Missing values in Promotion DataFrame:
code            0
supermarkets    0
week            0
feature         0
display         0
province        0
dtype: int64

Missing values in Supermarkets DataFrame:
supermarket_No    0
postal-code       0
dtype: int64


In [9]:
# Check for duplicate values
for df_name, df in zip(['Sales', 'Items', 'Promotion', 'Supermarkets'], 
                       [sales_df, items_df, promotion_df, supermarkets_df]):
    print(f"\Duplicate values in {df_name} DataFrame:")
    print(df.duplicated().sum())

\Duplicate values in Sales DataFrame:
0
\Duplicate values in Items DataFrame:
0
\Duplicate values in Promotion DataFrame:
0
\Duplicate values in Supermarkets DataFrame:
0


In [10]:
# Get a statistical summary
for df_name, df in zip(['Sales', 'Items', 'Promotion', 'Supermarkets'], 
                       [sales_df, items_df, promotion_df, supermarkets_df]):
    print(f"\nStatistical summary of {df_name} DataFrame:")
    print(df.describe())


Statistical summary of Sales DataFrame:
               code        amount         units          time      province  \
count  1.048575e+06  1.048575e+06  1.048575e+06  1.048575e+06  1.048575e+06   
mean   6.067646e+09  1.780470e+00  1.188219e+00  1.543165e+03  1.441374e+00   
std    3.154184e+09  5.966503e+00  5.466197e-01  3.833751e+02  4.965514e-01   
min    1.111124e+08 -8.280000e+00  1.000000e+00  0.000000e+00  1.000000e+00   
25%    3.620000e+09  9.900000e-01  1.000000e+00  1.259000e+03  1.000000e+00   
50%    5.100003e+09  1.500000e+00  1.000000e+00  1.601000e+03  1.000000e+00   
75%    9.999982e+09  2.190000e+00  1.000000e+00  1.824000e+03  2.000000e+00   
max    9.999986e+09  5.900000e+03  1.000000e+02  2.359000e+03  2.000000e+00   

               week    customerId   supermarket        basket           day  \
count  1.048575e+06  1.048575e+06  1.048575e+06  1.048575e+06  1.048575e+06   
mean   1.347071e+01  2.003481e+05  1.980431e+02  3.333491e+05  9.119343e+01   
std    8.5

Observations:

-There are no missing values, duplicates in any of the datasets, which is good for analysis.

-The Sales DataFrame contains transaction-level data, with each row likely representing a single purchase.

-The Items DataFrame provides details about each product.

-The Promotion DataFrame seems to contain information about promotional activities for specific products in specific supermarkets and weeks.

-The Supermarkets DataFrame provides location information for each supermarket.

-'amount' ranges from -8.28 to 5900.0, with a mean of 1.78.

-'time' ranges from 0 to 2359, likely representing time of day.

-The difference in 'week' ranges between Sales and Promotion DataFrames needs to be addressed.

Data Preprocessing

In [11]:
#Handle negative values in 'amount' column
print(f"Negative amounts before: {(sales_df['amount'] < 0).sum()}")
sales_df['amount'] = sales_df['amount'].abs()
print(f"Negative amounts after: {(sales_df['amount'] < 0).sum()}")

Negative amounts before: 1151
Negative amounts after: 0


In [12]:
# Investigate the 'time' column
print("Unique values in 'time' column:")
print(sales_df['time'].unique())
print("\nValue counts of 'time' column:")
print(sales_df['time'].value_counts().head())
print("\nDatatype of 'time' column:", sales_df['time'].dtype)


Unique values in 'time' column:
[1100 1137 1148 ...  531  328  438]

Value counts of 'time' column:
time
1738    2027
1734    2025
1718    2025
1713    2023
1745    2020
Name: count, dtype: int64

Datatype of 'time' column: int64


In [13]:
# Convert 'time' to a standard time format
def convert_time(time_value):
    time_str = f"{time_value:04d}"  # Pad with leading zeros if necessary
    return f"{time_str[:2]}:{time_str[2:]}"

sales_df['time'] = sales_df['time'].apply(convert_time)
sales_df['time'] = pd.to_datetime(sales_df['time'], format='%H:%M').dt.time

In [14]:
print("\nConverted 'time' column:")
print(sales_df['time'].head())
print("\nUnique values in converted 'time' column:")
print(sales_df['time'].nunique())


Converted 'time' column:
0    11:00:00
1    11:00:00
2    11:37:00
3    11:48:00
4    13:23:00
Name: time, dtype: object

Unique values in converted 'time' column:
1440


In [15]:
sales_df.head(10)

code  amount  units      time  province  week  customerId  \
0  7680850106    0.80      1  11:00:00         2     1      125434   
1  3620000470    3.59      1  11:00:00         2     1      125434   
2  1800028064    2.25      1  11:37:00         2     1      108320   
3  9999985067    0.85      1  11:48:00         2     1      162016   
4  9999985131    2.19      1  13:23:00         2     1       89437   
5  5100002794    2.19      1  13:23:00         2     1       89437   
6  1800000957    3.45      1  14:15:00         2     1      158549   
7  9999985051    1.29      1  14:15:00         2     1      158549   
8  9999985053    0.75      1  15:13:00         2     1       18851   
9  3620000446    2.19      1  15:23:00         2     1      118337   

   supermarket  basket  day  voucher  
0          244       1    1        0  
1          244       1    1        0  
2          244       2    1        0  
3          244       3    1        0  
4          244       4    1        0  
5          244       4    1        0  
6          244       5    1        0  
7          244       5    1        0  
8          244       6    1        0  
9          244       7    1        0

In [16]:
#handle week 
# Print original week ranges
print("Original week range in Sales DataFrame:", sales_df['week'].min(), "-", sales_df['week'].max())
print("Original week range in Promotion DataFrame:", promotion_df['week'].min(), "-", promotion_df['week'].max())


Original week range in Sales DataFrame: 1 - 28
Original week range in Promotion DataFrame: 43 - 104


In [17]:
# Align weeks
min_week_promo = promotion_df['week'].min()
sales_df['week'] = sales_df['week'] + min_week_promo - 1

In [18]:
# Print new week range in Sales DataFrame
print("New week range in Sales DataFrame:", sales_df['week'].min(), "-", sales_df['week'].max())


New week range in Sales DataFrame: 43 - 70


In [19]:
# Verify alignment
print("\nDo the ranges now align?")
print("Minimum week matches:", sales_df['week'].min() == promotion_df['week'].min())
print("Maximum week in Sales is within Promotion range:", sales_df['week'].max() <= promotion_df['week'].max())


Do the ranges now align?
Minimum week matches: True
Maximum week in Sales is within Promotion range: True
